### PR TITLE
Coefficient of Determination (R²) Metric Document

### PR DESCRIPTION
**Pull request by sdhossain**
### Linked issue(s)

[KOL-5759](https://linear.app/kolena/issue/KOL-5759/metrics-guide-for-r2)

### What change does this PR introduce and why?

Metrics documentation for R^2.

### Please check if the PR fulfills these requirements

- [x] Include reference to internal ticket and/or GitHub issue "Fixes #NNNN" (if applicable)
- [x] Relevant tests for the changes have been added
- [x] Relevant docs have been added / updated

### CODE REVIEW

File: docs/metrics/coefficient-of-determination.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
+# Coefficient of Determination (R²)
+
+The Coefficient of Determination, commonly denoted as R², is a statistical measure used to assess the goodness of fit
+of a regression model. It represents the proportion of the variance in the dependent variable that is predictable from
+the independent variables.
```
Comment (y27choi) [inline docs/metrics/coefficient-of-determination.md]: Does "dependent variable" here mean "inference" and "independent variables" mean "ground truth"? I wonder if we should use the terminologies that we commonly use.

File: docs/metrics/coefficient-of-determination.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+R² provides a scale that is **generally** from 0 to 1, where higher values indicate a better fit and imply that
+the model can better explain the variation of the output with the input variables. It is particularly useful for
```
Comment (y27choi) [inline docs/metrics/coefficient-of-determination.md]: suggest using `ground truth` and `inference` instead.

File: docs/metrics/coefficient-of-determination.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+metrics offer clearer insights.
+- High R² values do not imply that changes in predictors causally affect the outcome variable; R² measures
+correlation strength, not causation.
+
+It is also important to note that when stratifying properties to see how the model performances differ in varying
+conditions, it is important to note the affect of those conditions on the mean of the ground truth - as it influences
+the R² value. This is why if you were to stratify by age in an age estimation problem, you may even get negative values
+depending on the granularity of the strata.
+
+Error metrics like Mean Absolute Error (MAE) and Mean Squared Error (MSE) enrich model evaluation. MAE provides the
```
Comment (y27choi) [inline docs/metrics/coefficient-of-determination.md]: I would link "MAE" and "MSE" docs here

File: docs/metrics/coefficient-of-determination.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+indicating the variance in the dependent variable explained by the model. However, its utility is nuanced:
+
+- R² might overly favor the model's performance if the unseen data's distribution diverges from the training data,
+hinting at potential data characteristic shifts rather than model overfitting.
+- It might not capture the full spectrum of model accuracy in complex, non-linear scenarios, where direct error
+metrics offer clearer insights.
+- High R² values do not imply that changes in predictors causally affect the outcome variable; R² measures
+correlation strength, not causation.
+
+It is also important to note that when stratifying properties to see how the model performances differ in varying
```
Comment (y27choi) [inline docs/metrics/coefficient-of-determination.md]: run on sentence? should we break this up into 2 sentences?

File: docs/metrics/coefficient-of-determination.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+\end{align}
+$$
+</div>
+
+## Limitations and Biases
+
+When assessing regression models on unseen data, R² quantifies how well model predictions align with observed outcomes
+indicating the variance in the dependent variable explained by the model. However, its utility is nuanced:
+
+- R² might overly favor the model's performance if the unseen data's distribution diverges from the training data,
```
Comment (y27choi) [inline docs/metrics/coefficient-of-determination.md]: What does it mean by overly favoring? Will R^2 be large when there is data characteristic shifts?

File: docs/metrics/index.md [hunk 6]
```diff
@@ -35,6 +35,13 @@ biases, and its intended uses.
     Different averaging methods for aggregating metrics for **multiclass** workflows, such as classification and
     object detection.
 
+- [Coefficient of Determination](coefficient-of-determination.md)
```
Comment (y27choi) [inline docs/metrics/index.md]: Let's add `R^2` to the header

File: docs/metrics/index.md [hunk 7]
```diff
@@ -35,6 +35,13 @@ biases, and its intended uses.
...
+
+    ---
+
+    R² provides a scale that is generally from 0 to 1, where higher values indicate a better fit and imply that
+    the model can better explain the variation of the output with the input variables.
```
Comment (y27choi) [inline docs/metrics/index.md]: Let's also update the terminologies here.

File: docs/metrics/coefficient-of-determination.md (hunk 1 above)
Comment (sdhossain) [reply docs/metrics/coefficient-of-determination.md]: Hmm - dependent variable is actually both the inference and ground truth. (more-so the inference, as that is what is outputted from the model) and the independent varialbes are what we feed into the models.

X = independent variables
y = dependent

Comment (sdhossain) [reply docs/metrics/coefficient-of-determination.md]: To explain it a bit more perhaps - (maybe i should add this in actually)

the formula basically is:
R² = 1 - ( Mean Squared Error / Variance )

the **total variance of the dependent variable is the denominator** and the **unexplained variance of the dependent variable** is the error which is MSE.

What R² does is quantify how much variance of this predicted variable is predictable/explainable from the independent variables (the inputs of the model).

Comment (sdhossain) [reply docs/metrics/coefficient-of-determination.md]: Should I use input/output instead?

File: docs/metrics/coefficient-of-determination.md (hunk 5 above)
Comment (sdhossain) [reply docs/metrics/coefficient-of-determination.md]: Oh sorry - will try to make it a bit clearer.

Basically if unseen data diverges from training data. (And the test-data is not representative of that shift - the R^2 value might give a bad fit). ~fitting the distribution well but not generalizing well (kind of similar to overfitting actually - so will change the wording),

File: docs/metrics/index.md (hunk 7 above)
Comment (sdhossain) [reply docs/metrics/index.md]: I've reworded it (also some discussion on this thread as per before to "model inferences" and "model inputs"

File: docs/metrics/coefficient-of-determination.md (hunk 4 above)
Comment (sdhossain) [reply docs/metrics/coefficient-of-determination.md]: reworded

File: docs/metrics/coefficient-of-determination.md (hunk 5 above)
Comment (sdhossain) [reply docs/metrics/coefficient-of-determination.md]: I've changed it but if its not of too much use - am open to removing it as well

File: docs/metrics/coefficient-of-determination.md (hunk 2 above)
Comment (sdhossain) [reply docs/metrics/coefficient-of-determination.md]: reworded (also discussed in previous thread)

File: docs/metrics/coefficient-of-determination.md (hunk 1 above)
Comment (y27choi) [reply docs/metrics/coefficient-of-determination.md]: Hmm the independent variables (inputs) are not actually getting used in the formula. I'm trying to wrap my head around what the inputs have anything to do with the metric. Could you help me understand this a bit more please?

File: docs/metrics/coefficient-of-determination.md (hunk 4 above)
Comment (y27choi) [reply docs/metrics/coefficient-of-determination.md]: Could you take another look at this sentence?  It's still a bit awk with repeating "it is important to note..."

File: docs/metrics/coefficient-of-determination.md (hunk 1 above)
Comment (sdhossain) [reply docs/metrics/coefficient-of-determination.md]: The inputs aren't used in the formula - but they are used in the regressor to get predictions. But I understand the confusion.

I've reworded this to say "variance explained by the model"

Comment (sdhossain) [reply docs/metrics/coefficient-of-determination.md]:
```
R² provides a scale that is generally from 0 to 1, where higher values indicate a better fit and imply that the model can better explain the variation of the model inferences.
 ```
 
 and 
 
 ```
 It represents the proportion of the variance in the model inferences that is
explainable by the model itself.
```

is this better?

File: docs/metrics/coefficient-of-determination.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
+# Coefficient of Determination (R²)
+
+The Coefficient of Determination, commonly denoted as R², is a statistical measure used to assess the goodness of fit
+of a regression model. It represents the proportion of the variance in the model inferences that is
+explainable by the model itself.
```
Comment (y27choi) [inline docs/metrics/coefficient-of-determination.md:5]: nit: should this be "explained"?

**Review by y27choi — APPROVED** [review]
Left a nit comment but other than that 👍

File: docs/metrics/coefficient-of-determination.md (hunk 8 above)
Comment (sdhossain) [reply docs/metrics/coefficient-of-determination.md:5]: I think both work - chose explainable so that it is kind of future-tense-y as well haha.
